libshaderc_util/include/libshaderc_util/mutex.h

```diff
@@ -0,0 +1,134 @@
+// Copyright 2015 The Shaderc Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSHADERC_UTIL_INC_MUTEX_H
+#define LIBSHADERC_UTIL_INC_MUTEX_H
+
+// shaderc_util::mutex will be defined and specialized
+// depending on the platform that is being compiled.
+// It is more or less conformant to the C++11 specification of std::mutex.
+// However it does not implement try_lock.
+
+#ifdef _WIN32
+// windows.h #defines min and max if we don't define this.
+// this means things like std::min and std::max break
+#define NOMINMAX
+#include <windows.h>
+namespace shaderc_util {
+
+// As the name suggests, this mutex class is for running on windows.
+// It conforms to the c++11 mutex implementation, and should be a
+// drop in replacement.
+class windows_mutex {
+ public:
+  using native_handle_type = HANDLE;
+
+  windows_mutex() {
+    _mutex = CreateMutex(nullptr, false, nullptr);
+  }
+
+  ~windows_mutex() {
+    if (_mutex != INVALID_HANDLE_VALUE) {
+      CloseHandle(_mutex);
+    }
+  }
+
+  windows_mutex(windows_mutex&& _other) {
+    _mutex = _other.mutex;
+    _other.mutex = INVALID_HANDLE_VALUE;
```

Comment:
What happens if two different threads do this concurrently on the same `_other`?